storage/remote/queue_manager.go

```diff
@@ -800,6 +806,35 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan sample) {
 			s.droppedOnHardShutdown.Add(uint32(droppedSamples))
 			return
 
+		case <-s.reshard:
+			for {
+				if s.qm.blockIncomingSamples.Load() {
+					// The incoming samples will now be blocked for enty into the new shards,
+					// so its sample to stop listening samples.
+					break
+				}
```

Comment:
I don't think it is safe to stop reading from the queue as soon as we block new enqueues. There will still be a potential backlog of samples to work through. I think just a for loop across all samples in `queue` and adding them to pendingSamples will do the trick.

Reply:
Yes, this is better. Thank you for pointing it.